File: com.sgapsmae.client/Runtime/PixelExtractor.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace SGAPSMAEClient
{
    /// <summary>
    /// Lightweight pixel extractor for game frames.
    /// Optimized for minimal CPU overhead.
    /// </summary>
    public class PixelExtractor
    {
        private readonly int _frameWidth;
        private readonly int _frameHeight;
        private readonly int _targetWidth;
        private readonly int _targetHeight;
        
        // Reusable buffers to avoid allocation
        private readonly List<Vector2Int> _validCoords;
        private readonly List<Color32> _pixelValues;
        
        /// <summary>
        /// Create a new pixel extractor.
        /// </summary>
        /// <param name="config">Client configuration</param>
        public PixelExtractor(ClientConfig config)
        {
            _frameWidth = config.frameWidth;
            _frameHeight = config.frameHeight;
            _targetWidth = config.targetWidth;
            _targetHeight = config.targetHeight;
            
            _validCoords = new List<Vector2Int>(1000);
            _pixelValues = new List<Color32>(1000);
        }
        
        /// <summary>
        /// Extract pixels from texture at specified coordinates.
        /// </summary>
        /// <param name="texture">Source texture</param>
        /// <param name="coordinates">Sampling coordinates in target resolution</param>
        /// <param name="validCoords">Output valid coordinates</param>
        /// <param name="pixelValues">Output pixel values</param>
        public void Extract(
            Texture2D texture, 
            Vector2Int[] coordinates,
            out Vector2Int[] validCoords,
            out Color32[] pixelValues)
        {
            _validCoords.Clear();
            _pixelValues.Clear();
            
            // Get raw pixel data
            var pixels = texture.GetPixels32();
            int texWidth = texture.width;
            int texHeight = texture.height;
            
            // Scale factors
            float scaleU = (float)texHeight / _targetHeight;
            float scaleV = (float)texWidth / _targetWidth;
            
            foreach (var coord in coordinates)
            {
                // Scale coordinate to texture space
                int u = Mathf.FloorToInt(coord.x * scaleU);
                int v = Mathf.FloorToInt(coord.y * scaleV);
                
                // Bounds check
                if (u >= 0 && u < texHeight && v >= 0 && v < texWidth)
                {
                    // Unity texture coordinates: bottom-left origin
                    int pixelIndex = (texHeight - 1 - u) * texWidth + v;
                    
                    if (pixelIndex >= 0 && pixelIndex < pixels.Length)
                    {
                        _validCoords.Add(coord);
                        _pixelValues.Add(pixels[pixelIndex]);
                    }
                }
            }
            
            validCoords = _validCoords.ToArray();
            pixelValues = _pixelValues.ToArray();
        }
        
        /// <summary>
        /// Fast extraction using NativeArray (for burst compilation).
        /// </summary>
        /// <param name="pixelData">Raw pixel data</param>
        /// <param name="textureWidth">Texture width</param>
        /// <param name="textureHeight">Texture height</param>
        /// <param name="coordinates">Sampling coordinates</param>
        /// <param name="validCoords">Output valid coordinates</param>
        /// <param name="pixelValues">Output pixel values</param>
        public void ExtractFast(
            Color32[] pixelData,
            int textureWidth,
            int textureHeight,
            Vector2Int[] coordinates,
            out Vector2Int[] validCoords,
            out Color32[] pixelValues)
        {
            _validCoords.Clear();
            _pixelValues.Clear();
            
            float scaleU = (float)textureHeight / _targetHeight;
            float scaleV = (float)textureWidth / _targetWidth;
            
            for (int i = 0; i < coordinates.Length; i++)
            {
                int u = (int)(coordinates[i].x * scaleU);
                int v = (int)(coordinates[i].y * scaleV);
                
                if (u >= 0 && u < textureHeight && v >= 0 && v < textureWidth)
                {
                    int pixelIndex = (textureHeight - 1 - u) * textureWidth + v;
                    
                    if (pixelIndex >= 0 && pixelIndex < pixelData.Length)
                    {
                        _validCoords.Add(coordinates[i]);
                        _pixelValues.Add(pixelData[pixelIndex]);
                    }
                }
            }
            
            validCoords = _validCoords.ToArray();
            pixelValues = _pixelValues.ToArray();
        }
    }
}

File: v5/client/pixel_extractor.py
"""
Pixel Extractor for SGAPS-MAE Client
Simple pixel sampling from game frames.
"""

from typing import List, Tuple, Optional

import numpy as np


class PixelExtractor:
    """
    Lightweight pixel extractor for game frames.
    Simply reads pixel values at specified coordinates.
    
    Designed to be minimal for low client CPU usage.
    """
    
    def __init__(
        self,
        frame_width: int = 256,
        frame_height: int = 240
    ):
        """
        Args:
            frame_width: Game frame width
            frame_height: Game frame height
        """
        self.frame_width = frame_width
        self.frame_height = frame_height
    
    def extract(
        self,
        frame: np.ndarray,
        coordinates: np.ndarray
    ) -> Tuple[np.ndarray, np.ndarray]:
        """
        Extract pixel values at specified coordinates.
        
        Args:
            frame: Game frame [H, W, 3] uint8
            coordinates: Sampling coordinates [N, 2] (u, v)
            
        Returns:
            valid_coords: Valid coordinates [M, 2]
            pixel_values: RGB values [M, 3] float32
        """
        valid_coords = []
        pixel_values = []
        
        for u, v in coordinates:
            u_int, v_int = int(u), int(v)
            
            # Bounds check
            if 0 <= u_int < frame.shape[0] and 0 <= v_int < frame.shape[1]:
                valid_coords.append([u_int, v_int])
                pixel_values.append(frame[u_int, v_int] / 255.0)
        
        return np.array(valid_coords), np.array(pixel_values, dtype=np.float32)
    
    def extract_fast(
        self,
        frame: np.ndarray,
        coordinates: np.ndarray
    ) -> Tuple[np.ndarray, np.ndarray]:
        """
        Fast vectorized pixel extraction.
        
        Args:
            frame: Game frame [H, W, 3] uint8
            coordinates: Sampling coordinates [N, 2] (u, v)
            
        Returns:
            valid_coords: Valid coordinates [M, 2]
            pixel_values: RGB values [M, 3] float32
        """
        coords_int = coordinates.astype(np.int32)
        
        # Bounds mask
        valid_mask = (
            (coords_int[:, 0] >= 0) & 
            (coords_int[:, 0] < frame.shape[0]) &
            (coords_int[:, 1] >= 0) & 
            (coords_int[:, 1] < frame.shape[1])
        )
        
        valid_coords = coords_int[valid_mask]
        
        # Extract pixels
        pixel_values = frame[valid_coords[:, 0], valid_coords[:, 1]] / 255.0
        
        return valid_coords, pixel_values.astype(np.float32)
    
    def scale_coordinates(
        self,
        coordinates: np.ndarray,
        target_height: int,
        target_width: int
    ) -> np.ndarray:
        """
        Scale coordinates from target resolution to frame resolution.
        
        Args:
            coordinates: Coordinates in target resolution [N, 2]
            target_height: Target height (e.g., 224)
            target_width: Target width (e.g., 224)
            
        Returns:
            Scaled coordinates [N, 2]
        """
        scale_u = self.frame_height / target_height
        scale_v = self.frame_width / target_width
        
        scaled = coordinates.copy().astype(np.float32)
        scaled[:, 0] *= scale_u
        scaled[:, 1] *= scale_v
        
        return scaled


class AdaptivePixelExtractor(PixelExtractor):
    """
    Adaptive pixel extractor with priority-based extraction.
    """
    
    def __init__(
        self,
        frame_width: int = 256,
        frame_height: int = 240,
        priority_threshold: float = 0.5
    ):
        """
        Args:
            frame_width: Game frame width
            frame_height: Game frame height
            priority_threshold: Threshold for priority extraction
        """
        super().__init__(frame_width, frame_height)
        self.priority_threshold = priority_threshold
    
    def extract_with_priority(
        self,
        frame: np.ndarray,
        coordinates: np.ndarray,
        priorities: np.ndarray
    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """
        Extract pixels with priority information.
        
        Args:
            frame: Game frame [H, W, 3] uint8
            coordinates: Sampling coordinates [N, 2]
            priorities: Priority values [N] float
            
        Returns:
            valid_coords: Valid coordinates [M, 2]
            pixel_values: RGB values [M, 3]
            valid_priorities: Priorities [M]
        """
        valid_coords, pixel_values = self.extract_fast(frame, coordinates)
        
        # Get priorities for valid coordinates
        coords_int = coordinates.astype(np.int32)
        valid_mask = (
            (coords_int[:, 0] >= 0) & 
            (coords_int[:, 0] < frame.shape[0]) &
            (coords_int[:, 1] >= 0) & 
            (coords_int[:, 1] < frame.shape[1])
        )
        valid_priorities = priorities[valid_mask]
        
        return valid_coords, pixel_values, valid_priorities
